### PR TITLE
Avoid yield in Bio.SeqIO SffIO

### PR DESCRIPTION
**Pull request by mdehoon**
The indexing code can also be cleaned up, but we'll get to that in a separate PR.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqIO/SffIO.py [hunk 1]
```diff
@@ -1030,112 +883,228 @@ def iterate(self, handle):
...
-        read_header_size = struct.calcsize(read_header_fmt)
-        read_flow_fmt = ">%iH" % number_of_flows_per_read
-        read_flow_size = struct.calcsize(read_flow_fmt)
-        assert 1 == struct.calcsize(">B")
-        assert 1 == struct.calcsize(">s")
-        assert 1 == struct.calcsize(">c")
-        assert read_header_size % 8 == 0  # Important for padding calc later!
-        # The spec allows for the index block to be before or even in the middle
-        # of the reads. We can check that if we keep track of our position
-        # in the file...
```
Comment (peterjc) [inline Bio/SeqIO/SffIO.py:1043]: I think this comment is worth preserving (now in the ``__next__``function where the code to skip over the index block is not explained).

File: Bio/SeqIO/SffIO.py [hunk 2]
```diff
@@ -643,155 +643,6 @@ def _sff_read_roche_index(handle):
 _valid_UAN_read_name = re.compile(r"^[a-zA-Z0-9]{14}$")
 
 
-def _sff_read_seq_record(
```
Comment (peterjc) [inline Bio/SeqIO/SffIO.py:646]: As I recall I had this as a top level self contained function for ease of development and debugging - but not essential.

**Comment by mdehoon** [comment]
Note that `_check_eof` gets called only if we try to iterate past the last record. This is to be consistent with the existing code. Alternatively, we could `_check_eof` when we finish parsing the last record, and thereby catch any errors earlier. They may not be noticed otherwise. For example, in this situation:
```python
>>> records = SeqIO.parse("somefilewithgarbageattheend", "sff")
>>> n = len(records)
>>> for i in range(n):
...     record = next(records)
```
the garbage at the end would never be noticed.

File: Bio/SeqIO/SffIO.py (hunk 2 above)
Comment (mdehoon) [reply Bio/SeqIO/SffIO.py:646]: We can clean this up when we look at the indexing code.

File: Bio/SeqIO/SffIO.py (hunk 1 above)
Comment (mdehoon) [reply Bio/SeqIO/SffIO.py:1043]: OK

**Comment by peterjc** [comment]
Catching any garbage at the end of an SFF file seems helpful.

**Comment by mdehoon** [comment]
> Catching any garbage at the end of an SFF file seems helpful.

I agree. The only problem with that is that then there will be no way to get the last record, as reading the last record will raise the ValueError signaling that there is garbage at the end of the file.

**Comment by peterjc** [comment]
I haven't thought about this from the current code, but surely you can yield the final record, and then on the next iteration either raise StopIteration if the file finishes properly, or raise ValueError if the file is invalid. That may be easier said than done ;)

**Comment by mdehoon** [comment]
> I haven't thought about this from the current code, but surely you can yield the final record, and then on the next iteration either raise StopIteration if the file finishes properly, or raise ValueError if the file is invalid. That may be easier said than done ;)

Actually that is what the code is currently doing (same as in Biopython 1.84).

**Pull request merged**